File: src/utils/rewrite_args.ts
import { invoke } from "@tauri-apps/api";
import { shiftArrary } from "./shift_array";
import { copy_files } from "./copy_files";

export default function rewrite_args(username: string, path: string) {
  invoke("generate_args_command")
    .then(() => {
      return invoke("read_args_command");
    })
    .then((res) => {
      let args: string[] = shiftArrary(res as string[], 2);

      let newVersion = false;
      let version = args.findIndex((e) => e === "--version");
      if (parseFloat(args[version + 1].replace(/\./g, "")) < 1181) {
        newVersion = true;
      }

      for (let i = 0; i < args.length; i++) {
        if (!newVersion && args[i].includes("-Djava.library.path")) {
          let old = args[i];
          args[i] = "-Djava.library.path=" + path;
          if (path) {
            copy_files(old.substring(20), path);
          }
        } else if (args[i].includes("--username")) {
          if (username) {
            args[i + 1] = username;
          } else {
            alert("No username specified");
          }
        }
      }
      return invoke("write_args_command", {
        args: args.join(" "),
        path: path,
      });
    })
    .then(() => {
      return invoke("launch_game_command");
    })
    .catch((error) => {
      console.error(`Error:\n${error}`);
      alert(`Error:\n${error}`);
    });
}

File: src-tauri/src/utils/rewrite_args.rs
use super::shift_array::shift_array;
use super::{
    copy_snapshot::copy_snapshot, generate_args::generate_args, launch_game::launch_game,
    read_arguments::read_args, write_arguments::write_args,
};

use std::error::Error;

fn rewrite_args(username: String, path: String) -> Result<(), Box<dyn Error>> {
    generate_args()?;
    let res = read_args()?;
    let mut args: Vec<String> = shift_array(res, 2);

    let mut new_version = false;
    let version_index = args.iter().position(|e| e == "--version").unwrap();
    if args[version_index + 1]
        .replace(".", "")
        .parse::<f64>()
        .unwrap()
        < 1181.0
    {
        new_version = true;
    }

    for i in 0..args.len() {
        if !new_version && args[i].contains("-Djava.library.path") {
            let old = args[i].clone();
            args[i] = format!("-Djava.library.path={}", path);
            if !path.is_empty() {
                copy_snapshot(&old[20..], &path)?;
            }
        } else if args[i].contains("--username") {
            if !username.is_empty() {
                args[i + 1] = username.clone();
            } else {
                eprintln!("no username specified");
            }
        }
    }
    write_args(args.join(" "))?;
    launch_game()?;

    Ok(())
}
